base/compiler/plugin.jl

```diff
@@ -0,0 +1,291 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+# plugin management
+# =================
+
+PluginContext(plugins::Vector{Any}) = _PluginContext{to_tuple_type(plugins)}
+
+get_plugins(@nospecialize(ctx::PluginContext)) = first(ctx.parameters).parameters::SimpleVector
+
+"""
+    (::Type{<:AbstractCompilerPlugin})(f) = f()
+
+The entry into a plugin context
+Enter a plugin context from the entry call `f()`.
+
+!!! warning
+    This functor shouldn't be overloaded, the whole system will be broken otherwise.
+"""
+(::Type{<:AbstractCompilerPlugin})(f) = f()
+# (::Type{<:AbstractCompilerPlugin})(f, args...; kwargs...) = f(args...; kwargs...)
+
+const PLUGIN_ENTRY_METHOD = first(methods(AbstractCompilerPlugin))
+
+# overdubbing
+# ===========
+
+(::PluginContext)(args...) = return # a dummy method body
+
+const NULL_CONTEXT  = PluginContext(Any[])
+const SHADOW_METHOD = let
+    ms = methods(NULL_CONTEXT, (Tuple{Any,Vararg{Any}}))
+    @assert length(ms) == 1
+    first(ms)
+end
+
+is_shadow((; linfo)::Union{InferenceState,OptimizationState}) = is_shadow(linfo)
+is_shadow((; def, specTypes)::MethodInstance) = isa(def, Method) && _is_shadow(def, specTypes)
+is_shadow((; method, spec_types)::MethodMatch) = _is_shadow(method, spec_types)
+function _is_shadow(def::Method, @nospecialize(spec_types))
+    def === SHADOW_METHOD || return false
+    ft = unwrap_unionall(spec_types).parameters[1]
+    return is_plugin_ft(ft)
```

Comment:
This seems an odd place to split the method as attempting dispatch (aka subtyping) here to determine whether to use the plugin would be unsound, as it must use type-intersection (aka `methods()`)

Reply:
Actual "method match" is done within `find_matching_methods`, which, when in a plugin context, first finds matching methods for the original type signature and then forcibly forms corresponding method matches against the shadow method, under the assumption the shadow method signature  `(::PluginContext)(args...)` can always match arbitrary signatures.

This `is_shadow` is a query to check if the already introduced `MethodInstance` or `MethodMatch` is from the shadow method and is formed from plugin context. The only purpose of this `is_plugin_ft` is that we need to exclude the case that `Union{}(args...)` matches the shadow method when not in a plugin context.